### PR TITLE
[SPARK-45917][PYTHON][SQL] Automatic registration of Python Data Source on startup

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to add the support of automatic Python Data Source registration.

**End user perspective:**

```bash
# Assume that `customsource` defined a short name as `custom` 
pip install pyspark_customsource
```

Users can directly use the Python Data Source

```python
df = spark.format("custom").load()
```

**Developer perspective:**

The packages should follow the structure below:
- The package name should start with `pyspark_` prefix
- `pyspark_*.DefaultSource` has to be defined that inherits `pyspark.sql.datasource.DataSource`

For example:

```
pyspark_customsource
├── __init__.py
 ...
```

`__init__.py`:

```python
from pyspark.sql.datasource import DataSource

class DefaultSource(Datasource):
    pass
```

### Why are the changes needed?

This allows the developers to release and maintain their 3rd party Python Data Sources separately (e.g., in PyPI), and end users can easily install the Python Data Source without doing anything other than just `pip install pyspark_their_source`

### Does this PR introduce _any_ user-facing change?

Yes, this allows users to `pip install pyspark_custom_source`, and automatically register it as Data Source available in Spark. 

### How was this patch tested?

Unittests were added.

Also manual test as below:

```bash
rm -fr pyspark_mysource
mkdir pyspark_mysource
cd pyspark_mysource
echo '
from pyspark.sql.datasource import DataSource, DataSourceReader, InputPartition

class TestDataSourceReader(DataSourceReader):
    def __init__(self, options):
        self.options = options
    def partitions(self):
        return [InputPartition(i) for i in range(3)]
    def read(self, partition):
        yield partition.value, str(partition.value)


class DefaultSource(DataSource):
    @classmethod
    def name(cls):
        return "test"
    def schema(self):
        return "x INT, y STRING"
    def reader(self, schema) -> "DataSourceReader":
        return TestDataSourceReader(self.options)
    @classmethod
    def name(cls):
        return "mysource"
' > __init__.py
cd ..
./bin/pyspark
```

```python
spark.read.format("mysource").load().show()
```

```
+---+---+
|  x|  y|
+---+---+
|  0|  0|
|  1|  1|
|  2|  2|
+---+---+
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by HyukjinKwon** [comment]
Oops, there's a bit of more fixes to make (although the basic cases work). Let me mark as a draft for now.

**Comment by HyukjinKwon** [comment]
Let me actually add the test cases here together while I am here.

File: core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala [hunk 1]
```diff
@@ -145,4 +149,48 @@ private[spark] object PythonUtils extends Logging {
...
+      // to build PySpark every time during development.
+      val sparkHome: String = {
+        require(
+          sys.props.contains("spark.test.home") || sys.env.contains("SPARK_HOME"),
+          "spark.test.home or SPARK_HOME is not set.")
+        sys.props.getOrElse("spark.test.home", sys.env("SPARK_HOME"))
+      }
+      val sourcePath = Paths.get(sparkHome, "python").toAbsolutePath
+      val py4jPath = Paths.get(
+        sparkHome, "python", "lib", PythonUtils.PY4J_ZIP_NAME).toAbsolutePath
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala:171]: We don't use it but we need it when we import .. e.g., https://github.com/apache/spark/blob/master/python/pyspark/__init__.py#L53 -> https://github.com/apache/spark/blob/master/python/pyspark/conf.py#L23

File: python/pyspark/sql/worker/lookup_data_sources.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        setup_spark_files(infile)
+        setup_broadcasts(infile)
+
+        _accumulatorRegistry.clear()
+
+        infos = {}
+        for info in iter_modules():
+            if info.name.startswith("pyspark_"):
+                mod = import_module(info.name)
+                if hasattr(mod, "DefaultSource") and isinstance(mod.DefaultSource, DataSource):
```
Comment (ueshin) [inline python/pyspark/sql/worker/lookup_data_sources.py:69]: `issubclass` instead of `isinstance`?

Comment (HyukjinKwon) [reply python/pyspark/sql/worker/lookup_data_sources.py:69]: argh

**Review by ueshin — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by panbingkun** [comment]
@HyukjinKwon 
I have reverted https://github.com/apache/spark/pull/44504 (CommitID: `229a4eaf547e5c263c749bd53f7f9a89f4a9bea9`). 
Based on the current running results, the `Run Spark on Kubernetes Integration test` failure of GA is related to this.

**Comment by HyukjinKwon** [comment]
Thx, let me fix up together at https://github.com/apache/spark/pull/44519

**Pull request closed without merging**